### ISSUE TITLE
Channels sometimes enter permanent flow control state

### ISSUE THREAD

**Issue by michaelklishin** [rabbitmq/rabbitmq-server#98]
We've [heard](https://groups.google.com/forum/#!topic/rabbitmq-users/zZbKRNv5_kM) from a user that channels in after upgrading to 3.5.0 enter flow control state and stay there. What you see on the list is in part a red herring.

A later report off-list suggests that

 * Channels are credit-blocked on pids on remote nodes (as suggested by the 1st pid segment)
 * Eliminating mirroring via a policy makes the issue go away

So this may be related to changes in bug26527.

**Comment by michaelklishin** [comment]
More feedback from the same user suggests it may happen even without mirroring.

**Comment by michaelklishin** [comment]
Nothing in the logs suggests there are any node failures or network partitions going on (sadly, I cannot post the log).

**Comment by michaelklishin** [comment]
Attempts to reproduce this with a 2 node cluster so were unsuccessful. the cluster in question has 8 nodes.

**Comment by michaelklishin** [comment]
More updates: the issue magically goes away after a period of time, sometimes 30 minutes, sometimes a few hours. We have ingress queue message rate charts for 3.4.4 vs. 3.5.0, will post them if allowed to do so.

**Comment by noahhaon** [comment]
Hi - I am the user that reported the issue initially.  Recently, we were experiencing rmq bug # 26527 with 3.4.4 which prompted the upgrade to 3.5.0.  

I have included the charts below, they reflect the positive increments to the `publish` counts for our queues over two timespans, one over the past 18 hours or so with 3.5.0 and one showing more typical behavior, and a 3 week period prior showing more typical load.  Please let me know if you wish to track any more stats.  We will likely downgrade to 3.4.4 on Monday.

3.5.0:
![image](https://cloud.githubusercontent.com/assets/170715/6993292/bdacc86c-daa4-11e4-957c-0709761f5e04.png)

3.4.4:
![image](https://cloud.githubusercontent.com/assets/170715/6993285/8f954cd8-daa4-11e4-9085-743873059b6d.png)

**Comment by noahhaon** [comment]
I would also note that the cluster is recovering from the downtime during the upgrade and exchange publish rates are very high, on the order of 20k/s - 30k/s.  Typical for us is around 10k/s - 15k/s

**Comment by michaelklishin** [comment]
@noahhaon thank you.

Can you plase post a few more charts for the same intervals, if available? 

 * Disk I/O
 * Delivery rates to consumers 
 * Egress network traffic (to consumers) 
 * CPU utilisation

**Comment by noahhaon** [comment]
I will need to check with our ops team to see if we were collecting I/O and CPU on the nodes during that period, unfortunately I think we may have experienced an outage on the local collectors during that time :/  

I don't think we ever had network traffic to consumers.

Queue ack rates:

3.5.0

![image](https://cloud.githubusercontent.com/assets/170715/6993543/9115c7ba-daac-11e4-9b84-4d1a390aa157.png)

3.4.4

![image](https://cloud.githubusercontent.com/assets/170715/6993535/4ad56c88-daac-11e4-98e7-84b1fcfd0a94.png)

**Comment by noahhaon** [comment]
Updated graphs

**Comment by michaelklishin** [comment]
Another data point: the cluster has HiPE enabled on Erlang 17.x.

**Comment by noahhaon** [comment]
We tried upgrading to 3.5.1 today and quickly ran into the same issue after a couple mirrored queues grew to around 3MM messages.

**Comment by noahhaon** [comment]
More details: this occurs with or without mirrored queues.  Publisher channels had publish confirmations enabled, and consumer channels had manual acknowledgements.

**Comment by noahhaon** [comment]
We can reliably reproduce this with HA queues on 3.5.1

**Comment by michaelklishin** [comment]
@noahhaon can you please provide some details about an environment we can use to reproduce?

Specifically:

 * Is 1 mirror sufficient?
 * What's network bandwidth? 
 * If a small producer and consumer(s) can be used to reproduce, can you please post the code?

I've been unsuccessful in reproducing this so far with 2 development machines.

**Comment by noahhaon** [comment]
We haven't had time to set up a different environment for this, but it happens reliably in our production 8 node environment with RMQ 3.5.1 and Erlang 17.3

* This happens with a policy of:
```
ha-mode:	exactly
ha-params:	2
```
* Currently idle with several million messages stuck in queues :(  Prior to this about ~10mbps between nodes
* We've yet to reproduce this under light load.  It only seems to happen under heavy load with several million messages in queues.  We have seen it with the HA policy applied prior to publishing, and seen it happen shortly after HA policy gets applied to running queues
* Attempting to remove the ha policy doesn't work for about half of the queues, the old policy remains applied.  

Please let me know if there is anything else I can provide?

**Comment by michaelklishin** [comment]
@noahhaon does it happen to you when the queues are mostly empty or only when they back up? Is it common for the queues to have millions of messages in your environment? Have you observed high disk I/O on any of the nodes?

**Comment by noahhaon** [comment]
* It is when they back up
* Yes, millions of messages in the queues is not unusual when recovering from downtime - which we've had plenty of recently due to RMQ upgrades/downgrades :(
* Not particularly high I/O according to the RMQ management plugin.  I/O is quiescent currently and the cluster is completely unresponsive.

I have emailed you the output of maybe_stuck from one of the nodes which had a process with a huge dictionary of ctags.

**Comment by michaelklishin** [comment]
One node cannot keep up with the other (e.g. due to paging out a lot of data to disk) and channel that may send it messages gets blocked (no idea why it may last for minutes, though). Previously it would simply result in a higher RAM use on one of the nodes.

We'll see if we can make mirroring flow control optional for `3.5.2` as I'm afraid it may take a while to reproduce.

**Comment by noahhaon** [comment]
Sent logs from all the nodes.  I suspect this is something related to pub confirms, but I can't say for sure.  Note that all publisher connections were closed when these logs were generated.  I tried closing the consumer connections, but it caused the cluster to become unresponsive.

**Comment by noahhaon** [comment]
This has lasted for 10 hours so far.  Is there any way to dump the contents of a queue via the erlang console?  We have 10s of millions of messages we are at risk of losing now.

**Comment by michaelklishin** [comment]
No, what has changed in 3.5.0 has little to do with publisher confirms.

**Comment by noahhaon** [comment]
I just saw the huge number of ctags in the process dictionaries in the maybe stuck logs and though it could be related.

**Comment by michaelklishin** [comment]
They are unacknowledged messages delivered to consumers, in the `cr` record.

**Comment by noahhaon** [comment]
Hmmm.. I tried closing the consumer connections via the management api, and the management plugins on all the nodes became, and remain unresponsive.

**Comment by noahhaon** [comment]
re: making mirror flow control optional, I don't see anything in the logs indicating that they are blocking?  We switched to 3.5.x for the mirror flow control feature, as we were regularly hitting high memory watermark on our mirror nodes.

**Comment by michaelklishin** [comment]
@noahhaon I understand that but this is the most likely change that may cause processes to become blocked. We've seen stuck processes with exhausted credit in the other report of something very similar.

**Comment by noahhaon** [comment]
I agree.  We are going to have to start on disaster recovery here soon.  Is there anything else you would like me to run on the cluster while it is in this state?

**Comment by michaelklishin** [comment]
@noahhaon full logs from all nodes, if possible. Thank you.

**Comment by noahhaon** [comment]
@michaelklishin sent - if you think of anything else you need, please let me know.

**Comment by michaelklishin** [comment]
I've shared a tarball with a fix for #114 (https://github.com/rabbitmq/rabbitmq-server/commit/0b91553bafde0e063714b9557a1a76dc5213163b) with @noahhaon. Waiting to see if disabling gm flow control makes things return back to normal.

**Comment by noahhaon** [comment]
Thanks - as this issue has caused data loss for us, we will try to get a separate environment set up to test with and may be a week or so before we are able to get back to you,

**Comment by RonTsai** [comment]
Hi,   Michaelklishin ,  This is Ron from VMware.
 My client have opened bug26527, and also open an issue like this.
I can easy to reproduce this issue with two node in one cluster with mirror mode.
{ha-mode:all ,  ha-sync-mode:automatic}
Here are the reproduce procedure on RMQ3.5.

1. First queued 1G messages on queue in Master node( each message 3Kb, total 300000 msgs)
sh runjava.sh com.rabbitmq.examples.PerfTest -y0 -s 3000 -r 2 -x1000  -C 3000 -e exchange -f persistent -b 11 -p —uri amqp://user:pwd@host:5672/fdc

2. Then use perf tool to do stress test.
  Rabbitmq setting: Net_ticktime 90s
#28MB/s,  producer on Master node
sh runjava.sh com.rabbitmq.examples.PerfTest -y0 -s 14000 -r 2 -x1000  -e exchange -f persistent -b 11 -p —uri amqp://user:pwd@masterhost:5672/fdc

#100 consumer in Master node
sh runjava.sh com.rabbitmq.examples.PerfTest -y100  -u f12a.fdc.topic.exchange -p —uri amqp://user:pwd@masterhost:5672/fdc

3. Here we can see it’s soon to consume 300000 msg within 60~70 seconds.

4. But after 10mins around , the connection start to have flow control / block publish, and
memory kept on around 3G/10G (master/slave)

In the same perftool setting with 28MB/s producer and 100 consumer, and without the stuck messages. It will be no flow control, and the master/slave memory are keep around 400MB/100MB .

Also I can reproduce this on RMQ 3.2.1.
this will cause memory increasing without flow control, and result in over memory watermark.
The difference between 3.5. It will have flow control step in while memory increased over 3G/10G (master/slave). 

I think it is the native issue fd RMQ.  under high throughput, the mirror node within cluster will have this issue.I have test the same procedure under 18MB/s throughput, it will keep memory as normal, and won't have flow control. 

By the way I also done test of queuing message from 300000 to 6000 msg, it all have this issue, but under 5000 msg, it keep normal on memory and no flow control.

 Thanks for your help.

**Comment by michaelklishin** [comment]
@RonTsai thanks. There will be a way to disable inter-node flow control until we can find out what the root cause is exactly. Yes, disabling it would mean that mirrors can fall back behind the master (the case with 3.2.1 you are mentioning).

**Comment by RonTsai** [comment]
Hi,  MK
Could you also provide the tarball with a fix to me?
My client also need this patch to test.
Thanks!
Ron

**Comment by michaelklishin** [comment]
@RonTsai I don't see any email on your GitHub profile. How do I contact you?

**Comment by michaelklishin** [comment]
#114 was merged into `stable` and will be in `3.5.2`. It will allow those affected to go to the `3.4.x` behaviour (no flow control between master and mirrors) if they choose to do so.

**Comment by RonTsai** [comment]
Hi, sorry to miss your response.
My email: rtsai@vmware.com
Thanks

**Comment by michaelklishin** [comment]
Uploaded a `3.5.1.114` build: https://drive.google.com/folderview?id=0B12DYRM8lqNodi1nYXoxUDVoSmc&usp=sharing.

You can disable inter-node flow control by setting `rabbit.mirroring_flow_control` to false in the config.

**Comment by michaelklishin** [comment]
I'll close this with an explanation.

As of 3.5.2, it is possible to disable inter-node flow control. It does cause flow control to kick in more often but the real issue was that flow control status reporting was very confusing: things were reported as "in flow" if they were in flow at any moment in the last 5 seconds. But the flow state is not permanent, it goes on and off many times per second under load.

Some real improvements (for throughput) here would be:

 * Collecting a % statistics for flow control, much like we have for consumer utilisation
 * Make flow control more fine grained
 * Make mirroring involve fewer nodes by switching to a different algorithm, e.g. a flavour of Paxos. The one we have makes re-delivery in case of node failures a lot easier to reason about but is really inefficient because it involves nodes that may not even host mirrors for the queue.

The 2nd one is quite hard. We are working on 1 and 3, hope to have noticeable improvements by 3.6.0.
